keras_nlp/models/albert/albert_backbone.py

```diff
@@ -0,0 +1,271 @@
+            [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+        "padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)
+        ),
+    }
+
+    # Randomly initialized ALBERT encoder
+    model = keras_nlp.models.AlbertBackbone(
+        vocabulary_size=30000,
+        num_layers=12,
+        num_heads=12,
+        num_groups=1,
+        num_layers_per_group=1,
+        embedding_dim=128,
+        hidden_dim=768,
+        intermediate_dim=3072,
+        max_sequence_length=12,
+    )
+    output = model(input_data)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_heads,
+        num_groups,
+        num_layers_per_group,
+        embedding_dim,
+        hidden_dim,
+        intermediate_dim,
+        dropout=0.0,
+        max_sequence_length=512,
+        num_segments=2,
+        **kwargs,
+    ):
+
+        # Index of classification token in the vocabulary
+        cls_token_index = 0
+        # Inputs
+        token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="token_ids"
+        )
+        segment_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="segment_ids"
+        )
+        padding_mask = keras.Input(
+            shape=(None,), dtype="int32", name="padding_mask"
+        )
+
+        # Embed tokens, positions, and segment ids.
+        token_embedding_layer = keras.layers.Embedding(
+            input_dim=vocabulary_size,
+            output_dim=embedding_dim,
+            embeddings_initializer=albert_kernel_initializer(),
+            name="token_embedding",
+        )
+        token_embedding = token_embedding_layer(token_id_input)
+        position_embedding = PositionEmbedding(
+            initializer=albert_kernel_initializer(),
+            sequence_length=max_sequence_length,
+            name="position_embedding",
+        )(token_embedding)
+        segment_embedding = keras.layers.Embedding(
+            input_dim=num_segments,
+            output_dim=embedding_dim,
+            embeddings_initializer=albert_kernel_initializer(),
+            name="segment_embedding",
+        )(segment_id_input)
+
+        # Sum, normalize and apply dropout to embeddings.
+        x = keras.layers.Add()(
+            (token_embedding, position_embedding, segment_embedding)
+        )
+        x = keras.layers.LayerNormalization(
+            name="embeddings_layer_norm",
+            axis=-1,
+            epsilon=1e-12,
+            dtype=tf.float32,
+        )(x)
+        x = keras.layers.Dropout(
+            dropout,
+            name="embeddings_dropout",
+        )(x)
+
+        # Project the embedding to `hidden_dim`.
+        x = keras.layers.Dense(
+            hidden_dim,
+            kernel_initializer=albert_kernel_initializer(),
+            name="embedding_projection",
+        )(x)
+
+        layer_idx = 0
+        for i in range(num_groups):
```

Comment:
Hmmm, is this what's happening 🤔?

They have an argument called "inner_group_num" (which I have renamed as `num_layers_per_group` because `inner_group_num` isn't exactly easy to understand). Check this notebook out; the graph is similar to what I have done. And the number of parameters is the same for both.

https://colab.research.google.com/drive/17m3ozVYBTuodsSRxb8ycn9g37i6q48gc?usp=sharing

Reply:
@jbischof, yeah. The alternative is to have a separate loop for this. Something like:

```
        group_layers = []
        for i in range(num_groups):
            transformer_layers = [
                TransformerEncoder(
                    num_heads=num_heads,
                    intermediate_dim=intermediate_dim,
                    activation=lambda x: keras.activations.gelu(
                        x, approximate=True
                    ),
                    dropout=dropout,
                    kernel_initializer=albert_kernel_initializer(),
                    name=f"group_{i}_transformer_layer_{j}",
                )
                for j in range(num_layers_per_group)
            ]
            group_layers.append(transformer_layers)



        for layer_idx in range(num_layers):
            group_idx = int(layer_idx / (num_layers / num_groups))
            transformer_layers = group_layers[group_idx]
            for transformer_layer in transformer_layers:
                x = transformer_layer(x, padding_mask=padding_mask)
```

Reply:
Interesting! I missed the `"inner_group_num"`, but that seems fairly separate? A single "layer" according to the `num_layers` argument is just a stack of our encoders right? I think my solution could still stand just slightly expanded... 

```python
def get_layer():
    # A "layer" in Albert terminology is actually any number of repeated attention and
    # feed-forward blocks, controller by the `inner_group_num` parameter.
    layer = keras.Sequential()
    for _ in range(inner_group_num):
        layer.add(TransformerEncoder(...))
    return layer

for group_idx in range(num_groups):
    # Create a single encoder block with shared weights for the group.
    shared_encoder = get_layer()
    # Split our total layers evenly among groups.
    layers_per_group = num_layers // num_groups
    # Any remainder layers go into the earlier groups.
    if group_idx < num_layers % num_groups:
        layers_per_group += 1
    # Apply our shared encoder block once per layer in the group.
    for _ in range(layers_per_group):
        x = shared_encoder(x, padding_mask=padding_mask)
```

Maybe I am still missing something though, let's sync tomorrow!

Reply:
Hmmm, not quite. There is a difference between the two:

```
>>> for group_idx in range(5):
...   layers_per_group = 12 // 5
...   if group_idx < 12 % 5:
...     layers_per_group += 1
...   for _ in range(layers_per_group):
...     print(group_idx, end=", ")
... 
0, 0, 0, 1, 1, 1, 2, 2, 3, 3, 4, 4,
```

```
>>> for layer_idx in range(12):
...   group_idx = int (layer_idx / (12 / 5))
...   print(group_idx, end=", ")
... 
0, 0, 0, 1, 1, 2, 2, 2, 3, 3, 4, 4,
```

https://github.com/huggingface/transformers/blob/35a7052b61579cfe8df1a059d4cd3359310ec2d1/src/transformers/models/albert/modeling_albert.py#L473-L486

Reply:
Thanks! This might be it?

```python
layer_idx = 0
for group_idx in range(num_groups):
    # Create a single block with shared weights for the group.
    shared_block = get_shared_block()
    # Apply the shared block until the fraction of total layer exceeds
    # the fractional boundary of the next group.
    while layer_idx / num_layers < (group_idx + 1) / num_groups:
        x = shared_block(x, padding_mask=padding_mask)
        layer_idx += 1
```

Reply:
Yep, this should work!  

Also, 
```
    while layer_idx / num_layers < (group_idx + 1) / num_groups:
        x = shared_block(x, padding_mask=padding_mask)
        layer_idx += 1
```
is same as what is on the PR currently, if we just shift `num_groups` to the other side:
`while int(layer_idx / (num_layers / num_groups)) == group_idx`.